File: goblob/src/main/java/com/cauchymop/goblob/ui/AndroidGameRepository.java
package com.cauchymop.goblob.ui;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import com.cauchymop.goblob.model.Analytics;
import com.cauchymop.goblob.model.AvatarManager;
import com.cauchymop.goblob.model.GameDatas;
import com.cauchymop.goblob.model.GameRepository;
import com.cauchymop.goblob.proto.PlayGameData;
import com.cauchymop.goblob.proto.PlayGameData.GameData;
import com.cauchymop.goblob.proto.PlayGameData.GameData.Phase;
import com.cauchymop.goblob.proto.PlayGameData.GameList;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.PendingResult;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.games.Games;
import com.google.android.gms.games.Player;
import com.google.android.gms.games.multiplayer.Multiplayer;
import com.google.android.gms.games.multiplayer.Participant;
import com.google.android.gms.games.multiplayer.realtime.RoomConfig;
import com.google.android.gms.games.multiplayer.turnbased.LoadMatchesResponse;
import com.google.android.gms.games.multiplayer.turnbased.OnTurnBasedMatchUpdateReceivedListener;
import com.google.android.gms.games.multiplayer.turnbased.TurnBasedMatch;
import com.google.android.gms.games.multiplayer.turnbased.TurnBasedMatchBuffer;
import com.google.android.gms.games.multiplayer.turnbased.TurnBasedMatchConfig;
import com.google.android.gms.games.multiplayer.turnbased.TurnBasedMultiplayer;
import com.google.common.base.Predicate;
import com.google.common.base.Strings;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Iterables;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.TextFormat;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Lazy;

import static com.google.android.gms.games.Games.TurnBasedMultiplayer;

/**
 * Class to persist games.
 */
@Singleton
public class AndroidGameRepository extends GameRepository implements OnTurnBasedMatchUpdateReceivedListener {

  private static final String TAG = AndroidGameRepository.class.getName();
  private static final int CACHE_CHANGED_MESSAGE = 1;
  private static final long CACHE_CHANGED_DELAY = 100;
  private static final String IGNORED_VALUE = "";

  private final SharedPreferences prefs;
  private final GoogleApiClient googleApiClient;

  private AvatarManager avatarManager;

  private final Handler cacheRefreshHandler = new CacheRefreshHandler(this);

  @Inject
  public AndroidGameRepository(SharedPreferences prefs, GameDatas gameDatas,
      GoogleApiClient googleApiClient, AvatarManager avatarManager, Analytics analytics,
      @Named("PlayerOneDefaultName") Lazy<String> playerOneDefaultName,
      @Named("PlayerTwoDefaultName") String playerTwoDefaultName) {
    super(analytics, playerOneDefaultName, playerTwoDefaultName, gameDatas);
    this.prefs = prefs;
    this.googleApiClient = googleApiClient;
    this.avatarManager = avatarManager;
    gameCache = loadGameList();
    loadLegacyLocalGame();
    fireGameListChanged();
  }

  @Override
  protected void forceCacheRefresh() {
    requestCacheRefresh(true);
  }
  @Override
  protected void requestCacheRefresh(boolean immediate) {
    log("CacheRefreshHandler forceCacheRefresh()");
    cacheRefreshHandler.removeMessages(CACHE_CHANGED_MESSAGE);
    if (immediate) {
      cacheRefreshHandler.handleMessage(cacheRefreshHandler.obtainMessage(CACHE_CHANGED_MESSAGE));
    } else {
      cacheRefreshHandler.sendEmptyMessageDelayed(CACHE_CHANGED_MESSAGE, CACHE_CHANGED_DELAY);
    }
  }

  private void persistCache() {
    SharedPreferences.Editor editor = prefs.edit();
    editor.putString(GAMES, TextFormat.printToString(gameCache));
    editor.apply();
  }

  @Override
  public boolean publishRemoteGameState(GameData gameData) {
    if (googleApiClient.isConnected()) {
      Log.d(TAG, "publishRemoteGameState: " + gameData);
      String turnParticipantId = gameDatas.getCurrentPlayer(gameData).getId();
      byte[] gameDataBytes = gameData.toByteArray();
      Log.d(TAG, "takeTurn " + turnParticipantId);
      TurnBasedMultiplayer.takeTurn(googleApiClient, gameData.getMatchId(), gameDataBytes, turnParticipantId);
      if (gameData.getPhase() == Phase.FINISHED) {
        TurnBasedMultiplayer.finishMatch(googleApiClient, gameData.getMatchId());
        fireGameSelected(gameData);
      }
      return true;
    } else {
      gameCache.putUnpublished(gameData.getMatchId(), IGNORED_VALUE);
      return false;
    }
  }

  @Override
  protected void log(String message) {
    Log.d(AndroidGameRepository.TAG, message);
  }

  private void loadLegacyLocalGame() {
    String gameDataString = prefs.getString(GAME_DATA, null);
    if (gameDataString == null) {
      return;
    }
    log("loadLegacyLocalGame");
    GameData.Builder gameDataBuilder = GameData.newBuilder();
    try {
      TextFormat.merge(gameDataString, gameDataBuilder);
    } catch (TextFormat.ParseException e) {
      Log.e(TAG, "Error parsing local GameData: " + e.getMessage());
    }
    GameData localGame = gameDataBuilder.build();
    saveToCache(localGame);
    prefs.edit().remove(GAME_DATA).apply();
  }

  private GameList.Builder loadGameList() {
    log("loadGameList");
    String gameListString = prefs.getString(GAMES, "");
    GameList.Builder gameListBuilder = GameList.newBuilder();
    try {
      TextFormat.merge(gameListString, gameListBuilder);
    } catch (TextFormat.ParseException e) {
      Log.e(TAG, "Error parsing local GameList: " + e.getMessage());
    }
    log("loadGameList: " + gameListBuilder.getGames().size() + " games loaded.");
    return gameListBuilder;
  }

  public void refreshRemoteGameListFromServer() {
    Log.d(TAG, "refreshRemoteGameListFromServer");
    final long requestId = System.currentTimeMillis();

    PendingResult<com.google.android.gms.games.multiplayer.turnbased.TurnBasedMultiplayer.LoadMatchesResult> matchListResult =
        TurnBasedMultiplayer.loadMatchesByStatus(googleApiClient,
            Multiplayer.SORT_ORDER_SOCIAL_AGGREGATION,
            new int[]{TurnBasedMatch.MATCH_TURN_STATUS_MY_TURN, TurnBasedMatch.MATCH_TURN_STATUS_THEIR_TURN});
    ResultCallback<TurnBasedMultiplayer.LoadMatchesResult> matchListResultCallBack =
        new ResultCallback<TurnBasedMultiplayer.LoadMatchesResult>() {
          @Override
          public void onResult(@NonNull TurnBasedMultiplayer.LoadMatchesResult loadMatchesResult) {
            Log.d(TAG, String.format("matchResult: requestId = %d, latency = %d ms", requestId, System.currentTimeMillis() - requestId));
            LoadMatchesResponse matches = loadMatchesResult.getMatches();

            ImmutableList<TurnBasedMatch> allMatches = ImmutableList.<TurnBasedMatch>builder()
                .addAll(denullify(matches.getMyTurnMatches()))
                .addAll(denullify(matches.getTheirTurnMatches()))
                .addAll(denullify(matches.getCompletedMatches()))
                .build();

            Set<GameData> games = new HashSet<>();
            for (TurnBasedMatch match : allMatches) {
              updateAvatars(match);
              GameData gameData = getGameData(match);
              if (gameData != null) {
                games.add(gameData);
              }
            }
            if (clearRemoteGamesIfAbsent(games)) {
              forceCacheRefresh();
            }
            for (GameData game : games) {
              saveToCache(game);
            }
          }
        };
    matchListResult.setResultCallback(matchListResultCallBack);
  }

  private boolean clearRemoteGamesIfAbsent(final Set<GameData> games) {
    return Iterables.removeIf(gameCache.getMutableGames().values(), new Predicate<GameData>() {
      @Override
      public boolean apply(GameData gameData) {
        return gameDatas.isRemoteGame(gameData) && !games.contains(gameData);
      }
    });
  }

  private void updateAvatars(TurnBasedMatch match) {
    for (Participant participant : match.getParticipants()) {
      Player player = participant.getPlayer();
      avatarManager.setAvatarUri(player.getDisplayName(), player.getIconImageUri());
    }
  }

  private Iterable<TurnBasedMatch> denullify(@Nullable TurnBasedMatchBuffer nullableIterable) {
    return nullableIterable == null ? ImmutableList.of() : nullableIterable;
  }

  public GameData getGameData(@NonNull TurnBasedMatch turnBasedMatch) {
    byte[] data = turnBasedMatch.getData();
    if (data == null) {
      // When a crash happens during game creation, the TurnBasedMatch contains a null game
      // that can't be recovered, and would prevent starting the app.
      return null;
    }

    handleMatchStatusComplete(turnBasedMatch);
    try {
      GameData gameData = GameData.parseFrom(data);
      gameData = handleBackwardCompatibility(turnBasedMatch, gameData);
      return gameData;
    } catch (InvalidProtocolBufferException exception) {
      throw new RuntimeException(exception);
    }
  }

  private void handleMatchStatusComplete(@NonNull TurnBasedMatch turnBasedMatch) {
    boolean myTurn = turnBasedMatch.getTurnStatus() == TurnBasedMatch.MATCH_TURN_STATUS_MY_TURN;
    if (myTurn && turnBasedMatch.getStatus() == TurnBasedMatch.MATCH_STATUS_COMPLETE) {
      TurnBasedMultiplayer.finishMatch(googleApiClient, turnBasedMatch.getMatchId());
    }
  }

  private GameData createNewGameData(TurnBasedMatch turnBasedMatch) {
    String myId = getMyId(turnBasedMatch);
    String opponentId = getOpponentId(turnBasedMatch);
    PlayGameData.GoPlayer blackPlayer = createGoPlayer(turnBasedMatch, myId, true);
    PlayGameData.GoPlayer whitePlayer = createGoPlayer(turnBasedMatch, opponentId, false);
    GameData gameData = gameDatas.createNewGameData(turnBasedMatch.getMatchId(),
        PlayGameData.GameType.REMOTE, blackPlayer, whitePlayer);
    analytics.gameCreated(gameData);

    commitGameChanges(gameData);
    return gameData;
  }

  @NonNull
  private GameData handleBackwardCompatibility(TurnBasedMatch turnBasedMatch,
      GameData initialGameData) {
    GameData.Builder gameData = initialGameData.toBuilder();

    // No players
    if (!gameData.getGameConfiguration().hasBlack() || !gameData.getGameConfiguration().hasWhite()) {
      String myId = getMyId(turnBasedMatch);
      String opponentId = getOpponentId(turnBasedMatch);
      Map<String, PlayGameData.GoPlayer> goPlayers = ImmutableMap.of(
          myId, createGoPlayer(turnBasedMatch, myId, true),
          opponentId, createGoPlayer(turnBasedMatch, opponentId, false));
      PlayGameData.GameConfiguration gameConfiguration = gameData.getGameConfiguration();

      PlayGameData.GoPlayer blackPlayer = goPlayers.get(gameConfiguration.getBlackId());
      PlayGameData.GoPlayer whitePlayer = goPlayers.get(gameConfiguration.getWhiteId());

      gameData.getGameConfigurationBuilder().setBlack(blackPlayer).setWhite(whitePlayer);
    }

    // No match Id
    if (Strings.isNullOrEmpty(gameData.getMatchId())) {
      gameData.setMatchId(turnBasedMatch.getMatchId());
    }

    // No phase (version < 2)
    if (gameData.getPhase() == Phase.UNKNOWN) {
      final Phase result;
      if (gameData.hasMatchEndStatus()) {
        if (gameData.getMatchEndStatus().getGameFinished()) {
          result = Phase.FINISHED;
        } else {
          result = Phase.DEAD_STONE_MARKING;
        }
      } else {
        result = Phase.IN_GAME;
      }
      gameData.setPhase(result);
    }

    // No turn (version < 2)
    if (!gameData.hasTurn()) {
      if (gameData.hasMatchEndStatus()) {
        gameData.setTurn(gameData.getMatchEndStatus().getTurn());
      } else {
        PlayGameData.Color currentTurn = gameDatas.computeInGameTurn(gameData.getGameConfiguration(), gameData.getMoveCount());
        gameData.setTurn(currentTurn);
      }
    }

    return fillLocalStates(turnBasedMatch, gameData).build();
  }

  private GameData.Builder fillLocalStates(TurnBasedMatch turnBasedMatch, GameData.Builder gameData) {
    boolean isMyTurn = turnBasedMatch.getTurnStatus() == TurnBasedMatch.MATCH_TURN_STATUS_MY_TURN;
    boolean turnIsBlack = gameData.getTurn() == PlayGameData.Color.BLACK;
    boolean iAmBlack = isMyTurn && turnIsBlack || (!isMyTurn && !turnIsBlack);
    PlayGameData.GameConfiguration.Builder gameConfiguration = gameData.getGameConfigurationBuilder();
    PlayGameData.GoPlayer.Builder blackPlayer = gameConfiguration.getBlackBuilder();
    PlayGameData.GoPlayer.Builder whitePlayer = gameConfiguration.getWhiteBuilder();
    blackPlayer.setIsLocal(iAmBlack);
    whitePlayer.setIsLocal(!iAmBlack);
    Log.d(TAG, String.format("black: %s", blackPlayer));
    Log.d(TAG, String.format("white %s", whitePlayer));
    return gameData;
  }

  @NonNull
  private String getOpponentId(TurnBasedMatch turnBasedMatch) {
    String myId = getMyId(turnBasedMatch);
    for (String participantId : turnBasedMatch.getParticipantIds()) {
      if (!participantId.equals(myId)) {
        return participantId;
      }
    }
    throw new RuntimeException("Our TurnBasedMatch should contain 2 players!");
  }

  private String getMyId(TurnBasedMatch turnBasedMatch) {
    return turnBasedMatch.getParticipantId(Games.Players.getCurrentPlayerId(googleApiClient));
  }

  private PlayGameData.GoPlayer createGoPlayer(TurnBasedMatch match, String participantId,
      boolean isLocal) {
    PlayGameData.GoPlayer goPlayer;
    Player player = match.getParticipant(participantId).getPlayer();
    goPlayer = gameDatas.createGamePlayer(participantId, player.getDisplayName(), isLocal);
    avatarManager.setAvatarUri(player.getDisplayName(), player.getIconImageUri());
    return goPlayer;
  }

  @Override
  public void onTurnBasedMatchReceived(TurnBasedMatch turnBasedMatch) {
    Log.d(TAG, "onTurnBasedMatchReceived");
    GameData gameData = getGameData(turnBasedMatch);
    if (gameData != null) {
      saveToCache(gameData);
    }
  }

  @Override
  public void onTurnBasedMatchRemoved(String matchId) {
    Log.d(TAG, "onTurnBasedMatchRemoved: " + matchId);
    removeFromCache(matchId);
  }

  public void handlePlayersSelected(Intent intent) {
    Log.d(TAG, "handlePlayersSelected");

    // get the invitee list
    final ArrayList<String> invitees = intent.getStringArrayListExtra(Games.EXTRA_PLAYER_IDS);
    Log.d(TAG, "Invitees: " + invitees);

    // get the automatch criteria
    Bundle autoMatchCriteria = null;
    int minAutoMatchPlayers = intent.getIntExtra(Multiplayer.EXTRA_MIN_AUTOMATCH_PLAYERS, 0);
    int maxAutoMatchPlayers = intent.getIntExtra(Multiplayer.EXTRA_MAX_AUTOMATCH_PLAYERS, 0);
    if (minAutoMatchPlayers > 0 || maxAutoMatchPlayers > 0) {
      autoMatchCriteria = RoomConfig.createAutoMatchCriteria(
          minAutoMatchPlayers, maxAutoMatchPlayers, 0);
      Log.d(TAG, "Automatch criteria: " + autoMatchCriteria);
    }

    // create game
    TurnBasedMatchConfig turnBasedMatchConfig = TurnBasedMatchConfig.builder()
        .addInvitedPlayers(invitees)
        .setVariant(TurnBasedMatch.MATCH_VARIANT_DEFAULT)
        .setAutoMatchCriteria(autoMatchCriteria).build();

    // kick the match off
    TurnBasedMultiplayer.createMatch(googleApiClient, turnBasedMatchConfig)
        .setResultCallback(new ResultCallback<TurnBasedMultiplayer.InitiateMatchResult>() {
          @Override
          public void onResult(
              @NonNull TurnBasedMultiplayer.InitiateMatchResult initiateMatchResult) {
            Log.d(TAG, "InitiateMatchResult " + initiateMatchResult);
            if (!initiateMatchResult.getStatus().isSuccess()) {
              return;
            }
            TurnBasedMatch turnBasedMatch = initiateMatchResult.getMatch();
            createNewGameData(turnBasedMatch);

            Log.d(TAG, "Game created...");
            selectGame(turnBasedMatch.getMatchId());
          }
        });
  }

  public void handleCheckMatchesResult(int responseCode, Intent intent) {
    Log.d(TAG, "handleCheckMatchesResult");
    refreshRemoteGameListFromServer();
    if (responseCode == Activity.RESULT_OK) {
      final TurnBasedMatch match = intent.getParcelableExtra(Multiplayer.EXTRA_TURN_BASED_MATCH);
      selectGame(match.getMatchId());
    }
  }

  private static class CacheRefreshHandler extends Handler {

    private final AndroidGameRepository androidGameRepository;

    public CacheRefreshHandler(AndroidGameRepository androidGameRepository) {
      this.androidGameRepository = androidGameRepository;
    }

    @Override
    public void handleMessage(Message msg) {
      Log.d(TAG, "CacheRefreshHandler handleMessage");
      androidGameRepository.persistCache();
      androidGameRepository.fireGameListChanged();
    }
  }
}

File: goblob/src/main/java/com/cauchymop/goblob/ui/AndroidGameRepository.kt
package com.cauchymop.goblob.ui

import android.app.Activity
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import com.cauchymop.goblob.model.*
import com.cauchymop.goblob.proto.PlayGameData
import com.cauchymop.goblob.proto.PlayGameData.GameData
import com.cauchymop.goblob.proto.PlayGameData.GameData.Phase
import com.cauchymop.goblob.proto.PlayGameData.GameList
import com.google.android.gms.games.Games
import com.google.android.gms.games.TurnBasedMultiplayerClient
import com.google.android.gms.games.multiplayer.Multiplayer
import com.google.android.gms.games.multiplayer.realtime.RoomConfig
import com.google.android.gms.games.multiplayer.turnbased.LoadMatchesResponse
import com.google.android.gms.games.multiplayer.turnbased.TurnBasedMatch
import com.google.android.gms.games.multiplayer.turnbased.TurnBasedMatchConfig
import com.google.android.gms.games.multiplayer.turnbased.TurnBasedMatchUpdateCallback
import com.google.common.base.Strings
import com.google.common.collect.ImmutableMap
import com.google.protobuf.InvalidProtocolBufferException
import com.google.protobuf.TextFormat
import dagger.Lazy
import java.util.*
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Provider
import javax.inject.Singleton

/**
 * Class to persist games.
 */

private const val GAME_DATA = "gameData"
private const val GAMES = "games"
private const val TAG = "AndroidGameRepository"
private const val IGNORED_VALUE = ""

@Singleton
class AndroidGameRepository @Inject
constructor(private val prefs: SharedPreferences, gameDatas: GameDatas,
            private val googleAccountManager: GoogleAccountManager,
            private val turnBasedClientProvider: Provider<TurnBasedMultiplayerClient>,
            private val avatarManager: AvatarManager, analytics: Analytics,
            @Named("PlayerOneDefaultName") playerOneDefaultName: Lazy<String>,
            @Named("PlayerTwoDefaultName") playerTwoDefaultName: String) : GameRepository(analytics, playerOneDefaultName, playerTwoDefaultName, gameDatas, gameCache = loadGameCache(prefs)) {

  init {
    loadLegacyLocalGame()
    fireGameListChanged()
    googleAccountManager.addAccountStateListener(object : AccountStateListener {
      override fun accountStateChanged(isSignInComplete: Boolean) {
        if (isSignInComplete) {
          initTurnBasedUpdateListeners()
        }
      }
    })
  }

  private fun initTurnBasedUpdateListeners() {
    val turnBasedClient = turnBasedClientProvider.get()
    turnBasedClient.registerTurnBasedMatchUpdateCallback(object : TurnBasedMatchUpdateCallback() {
      override fun onTurnBasedMatchReceived(turnBasedMatch: TurnBasedMatch) {
        Log.d(TAG, "onTurnBasedMatchReceived")
        val gameData = getGameData(turnBasedMatch)
        if (gameData != null) {
          if (saveToCache(gameData)) {
            forceCacheRefresh()
          }
        }
      }

      override fun onTurnBasedMatchRemoved(matchId: String) {
        Log.d(TAG, "onTurnBasedMatchRemoved: $matchId")
        removeFromCache(matchId)
      }
    })
  }

  override fun forceCacheRefresh() {
    Log.d(TAG, "forceCacheRefresh")
    persistCache()
    fireGameListChanged()
  }


  private fun persistCache() {
    val editor = prefs.edit()
    editor.putString(GAMES, TextFormat.printToString(gameCache))
    editor.apply()
  }

  override fun publishRemoteGameState(gameData: GameData): Boolean {
    if (googleAccountManager.signInComplete) {
      Log.d(TAG, "publishRemoteGameState: $gameData")
      val turnParticipantId = gameDatas.getCurrentPlayer(gameData).id
      val gameDataBytes = gameData.toByteArray()
      Log.d(TAG, "takeTurn $turnParticipantId")
      val turnBasedClient = turnBasedClientProvider.get()
      turnBasedClient.takeTurn(gameData.matchId, gameDataBytes, turnParticipantId)
      if (gameData.phase == Phase.FINISHED) {
        turnBasedClient.finishMatch(gameData.matchId)
        fireGameSelected(gameData)
      }
      return true
    } else {
      gameCache.putUnpublished(gameData.matchId, IGNORED_VALUE)
      return false
    }
  }

  override fun log(message: String) {
    Log.d(TAG, message)
  }

  private fun loadLegacyLocalGame() {
    val gameDataString = prefs.getString(GAME_DATA, null) ?: return
    log("loadLegacyLocalGame")
    val gameDataBuilder = GameData.newBuilder()
    try {
      TextFormat.merge(gameDataString, gameDataBuilder)
    } catch (e: TextFormat.ParseException) {
      Log.e(TAG, "Error parsing local GameData: ${e.message}")
    }

    val localGame = gameDataBuilder.build()
    if (saveToCache(localGame)) {
      forceCacheRefresh()
    }
    prefs.edit().remove(GAME_DATA).apply()
  }

  fun refreshRemoteGameListFromServer() {
    Log.d(TAG, "refreshRemoteGameListFromServer -  currentMatchId = $currentMatchId")
    val requestId = System.currentTimeMillis()

    val turnBasedClient = turnBasedClientProvider.get()
    val matchListResult = turnBasedClient.loadMatchesByStatus(Multiplayer.SORT_ORDER_SOCIAL_AGGREGATION,
        intArrayOf(TurnBasedMatch.MATCH_TURN_STATUS_MY_TURN, TurnBasedMatch.MATCH_TURN_STATUS_THEIR_TURN, TurnBasedMatch.MATCH_TURN_STATUS_COMPLETE))
    matchListResult.addOnCompleteListener { task ->
      task.result?.get()?.let { matchResponseCallback(requestId, it) }

    }
  }

  private fun matchResponseCallback(requestId: Long, loadMatchesResponse: LoadMatchesResponse) {
    Log.d(TAG, String.format("matchResult: requestId = %d, latency = %d ms", requestId, System.currentTimeMillis() - requestId))
    val allMatches = with(loadMatchesResponse) { myTurnMatches.asSequence() + theirTurnMatches.asSequence() + completedMatches.asSequence() }
    val games = HashSet<GameData>()
    for (match in allMatches) {
      updateAvatars(match)
      val gameData = getGameData(match)
      if (gameData != null) {
        games.add(gameData)
      }
    }

    val removedMatchIds = clearRemoteGamesIfAbsent(games)
    val selectedIsGone = removedMatchIds.contains(currentMatchId)
    val changedCount = removedMatchIds.size + games.filter { saveToCache(it) }.count()
    if (changedCount > 0) {
      forceCacheRefresh()
    }

    // Select invitation if one has arrived
    invitationMatchId?.let {
      invitationMatchId = null
      selectGame(it)
    }

    Log.d(TAG, " ===> (in refreshRemoteGameListFromServer) changedCount is $changedCount and selectedIsGone is $selectedIsGone")
    if (selectedIsGone) {
      // selected game was removed, we select new game instead
      selectGame(NO_MATCH_ID)
    }

    loadMatchesResponse.release()
  }

  private fun clearRemoteGamesIfAbsent(games: Set<GameData>): List<String> {
    val keysToRemove = gameCache.gamesMap.filter { gameDatas.isRemoteGame(it.value) && !games.contains(it.value) }.map { it.key }
    keysToRemove.forEach { gameCache.removeGames(it) }
    return keysToRemove
  }

  private fun updateAvatars(match: TurnBasedMatch) {
    for (participant in match.participants) {
      val player = participant.player
      avatarManager.setAvatarUri(player.displayName, player.iconImageUri)
    }
  }

  private fun getGameData(turnBasedMatch: TurnBasedMatch): GameData? {
    val data = turnBasedMatch.data
        ?: // When a crash happens during game creation, the TurnBasedMatch contains a null game
        // that can't be recovered, and would prevent starting the app.
        return null

    handleMatchStatusComplete(turnBasedMatch)
    try {
      var gameData: GameData = GameData.parseFrom(data)
      gameData = handleBackwardCompatibility(turnBasedMatch, gameData)
      return gameData
    } catch (exception: InvalidProtocolBufferException) {
      throw RuntimeException(exception)
    }

  }

  private fun handleMatchStatusComplete(turnBasedMatch: TurnBasedMatch) {
    val myTurn = turnBasedMatch.turnStatus == TurnBasedMatch.MATCH_TURN_STATUS_MY_TURN
    if (myTurn && turnBasedMatch.status == TurnBasedMatch.MATCH_STATUS_COMPLETE) {
      val turnBasedClient = turnBasedClientProvider.get()
      turnBasedClient.finishMatch(turnBasedMatch.matchId)
    }
  }

  private fun createNewGameData(turnBasedMatch: TurnBasedMatch): GameData {
    val myId = getMyId(turnBasedMatch)
    val opponentId = getOpponentId(turnBasedMatch)
    val blackPlayer = createGoPlayer(turnBasedMatch, myId, true)
    val whitePlayer = createGoPlayer(turnBasedMatch, opponentId, false)
    val gameData = gameDatas.createNewGameData(turnBasedMatch.matchId,
        PlayGameData.GameType.REMOTE, blackPlayer, whitePlayer)
    analytics.gameCreated(gameData)

    commitGameChanges(gameData)
    return gameData
  }

  @Suppress("DEPRECATION")
  private fun handleBackwardCompatibility(turnBasedMatch: TurnBasedMatch,
                                          initialGameData: GameData): GameData {
    val gameData = initialGameData.toBuilder()

    // No players
    if (!gameData.gameConfiguration.hasBlack() || !gameData.gameConfiguration.hasWhite()) {
      val myId = getMyId(turnBasedMatch)
      val opponentId = getOpponentId(turnBasedMatch)
      val goPlayers = ImmutableMap.of(
          myId, createGoPlayer(turnBasedMatch, myId, true),
          opponentId, createGoPlayer(turnBasedMatch, opponentId, false))
      val gameConfiguration = gameData.gameConfiguration

      val blackPlayer = goPlayers[gameConfiguration.blackId]
      val whitePlayer = goPlayers[gameConfiguration.whiteId]

      gameData.gameConfigurationBuilder
          .setBlack(blackPlayer)
          .setWhite(whitePlayer)
    }

    // No match Id
    if (Strings.isNullOrEmpty(gameData.matchId)) {
      gameData.matchId = turnBasedMatch.matchId
    }

    // No phase (version < 2)
    if (gameData.phase == Phase.UNKNOWN) {
      gameData.phase = if (gameData.hasMatchEndStatus()) {
        if (gameData.matchEndStatus.gameFinished) {
          Phase.FINISHED
        } else {
          Phase.DEAD_STONE_MARKING
        }
      } else {
        Phase.IN_GAME
      }
    }

    // No turn (version < 2)
    if (!gameData.hasTurn()) {
      if (gameData.hasMatchEndStatus()) {
        gameData.turn = gameData.matchEndStatus.turn
      } else {
        val currentTurn = gameDatas.computeInGameTurn(gameData.gameConfiguration, gameData.moveCount)
        gameData.turn = currentTurn
      }
    }

    return fillLocalStates(turnBasedMatch, gameData).build()
  }

  private fun fillLocalStates(turnBasedMatch: TurnBasedMatch, gameData: GameData.Builder): GameData.Builder {
    val isMyTurn = turnBasedMatch.turnStatus == TurnBasedMatch.MATCH_TURN_STATUS_MY_TURN
    val turnIsBlack = gameData.turn == PlayGameData.Color.BLACK
    val iAmBlack = isMyTurn && turnIsBlack || !isMyTurn && !turnIsBlack
    val gameConfiguration = gameData.gameConfigurationBuilder
    val blackPlayer = gameConfiguration.blackBuilder
    val whitePlayer = gameConfiguration.whiteBuilder
    blackPlayer.isLocal = iAmBlack
    whitePlayer.isLocal = !iAmBlack
    Log.d(TAG, String.format("black: %s", blackPlayer))
    Log.d(TAG, String.format("white %s", whitePlayer))
    return gameData
  }

  private fun getOpponentId(turnBasedMatch: TurnBasedMatch): String {
    val myId = getMyId(turnBasedMatch)
    for (participantId in turnBasedMatch.participantIds) {
      if (participantId != myId) {
        return participantId
      }
    }
    throw RuntimeException("Our TurnBasedMatch should contain 2 players!")
  }

  private fun getMyId(turnBasedMatch: TurnBasedMatch) =
      turnBasedMatch.getParticipantId(googleAccountManager.currentPlayerId)

  private fun createGoPlayer(match: TurnBasedMatch, participantId: String,
                             isLocal: Boolean): PlayGameData.GoPlayer {
    val goPlayer: PlayGameData.GoPlayer
    val player = match.getParticipant(participantId).player
    goPlayer = gameDatas.createGamePlayer(participantId, player.displayName, isLocal)
    avatarManager.setAvatarUri(player.displayName, player.iconImageUri)
    return goPlayer
  }

  fun handlePlayersSelected(intent: Intent) {
    Log.d(TAG, "handlePlayersSelected")

    // get the invitee list
    val invitees = intent.getStringArrayListExtra(Games.EXTRA_PLAYER_IDS)
    Log.d(TAG, "Invitees: $invitees")

    // get the automatch criteria
    var autoMatchCriteria: Bundle? = null
    val minAutoMatchPlayers = intent.getIntExtra(Multiplayer.EXTRA_MIN_AUTOMATCH_PLAYERS, 0)
    val maxAutoMatchPlayers = intent.getIntExtra(Multiplayer.EXTRA_MAX_AUTOMATCH_PLAYERS, 0)
    if (minAutoMatchPlayers > 0 || maxAutoMatchPlayers > 0) {
      autoMatchCriteria = RoomConfig.createAutoMatchCriteria(
          minAutoMatchPlayers, maxAutoMatchPlayers, 0)
      Log.d(TAG, "Automatch criteria: " + autoMatchCriteria!!)
    }

    // create game
    val turnBasedMatchConfig = TurnBasedMatchConfig.builder()
        .addInvitedPlayers(invitees)
        .setVariant(TurnBasedMatch.MATCH_VARIANT_DEFAULT)
        .setAutoMatchCriteria(autoMatchCriteria).build()

    // kick the match off
    val turnBasedClient = turnBasedClientProvider.get()
    turnBasedClient.createMatch(turnBasedMatchConfig).addOnCompleteListener { initiateMatchResult ->
      Log.d(TAG, "InitiateMatchResult $initiateMatchResult")
      if (!initiateMatchResult.isSuccessful) {
        return@addOnCompleteListener
      }
      initiateMatchResult.result?.let {
        createNewGameData(it)

        Log.d(TAG, "Game created...")
        selectGame(it.matchId)
      }
    }
  }

  fun handleCheckMatchesResult(responseCode: Int, intent: Intent?) {
    Log.d(TAG, "handleCheckMatchesResult")
    // Refresh is done in all cases as games can be dismissed for the Check Matches Screen if we then
    // do not select a game and press back
    refreshRemoteGameListFromServer()

    if (responseCode == Activity.RESULT_OK && intent != null) {
      val match = intent.getParcelableExtra<TurnBasedMatch>(Multiplayer.EXTRA_TURN_BASED_MATCH)
      selectGame(match.matchId)
    }
  }

}

private fun loadGameCache(sharedPreferences: SharedPreferences): GameList.Builder {
  Log.d(TAG, "loadGameList")
  val gameListString = sharedPreferences.getString(GAMES, "")
  val gameListBuilder = GameList.newBuilder()
  try {
    TextFormat.merge(gameListString, gameListBuilder)
  } catch (e: TextFormat.ParseException) {
    Log.e(TAG, "Error parsing local GameList: " + e.message)
  }

  Log.d(TAG, "loadGameList: " + gameListBuilder.gamesCount + " games loaded.")
  return gameListBuilder
}
